shared/lib-hoppy/hoppy/mq.py

```diff
@@ -0,0 +1,117 @@
+import functools
+import json
+import logging
+import time
+
+import pika
+from hoppy.config import RABBITMQ_CONFIG
+
+
+class Service:
+    def __init__(self, exchange, consumers, config={}):
+        self.config = {**RABBITMQ_CONFIG, **config}
+        self.exchange = exchange
+        self.consumers = consumers
+
+        credentials = pika.PlainCredentials(config["username"], config["password"])
+        self.params = pika.ConnectionParameters(config["host"], config["port"], credentials=credentials)
+
+    def _connect(self):
+        for i in range(self.config["retry_limit"]):
+            try:
+                return pika.BlockingConnection(self.params)
+            except Exception as e:
+                logging.warning(e, exc_info=True)
+                logging.warning(f"RabbitMQ Connection Failed. Retrying in 30s ({i + 1}/{self.config['retry_limit']})")
+                time.sleep(30)
+        return None
+
+    def _setup_channel(self, connection):
+        channel = connection.channel()
+        channel.exchange_declare(exchange=self.exchange, exchange_type="direct", durable=True, auto_delete=True)
+        for name, callback in self.consumers.items():
+            QueueConsumer(name, self.exchange, callback).bind_to_channel(channel)
+            logging.info(f" [*] Waiting for data for queue: {name}. To exit press CTRL+C")
+        return channel
+
+    # When run, it attempts to create a pika.BlockingConnection() with the settings in CONSUMER_CONFIG
+    # There are 2 retry levels for the consumer. The first being in _create_connection() which is based on retry_limit
+    # The other being when the app crashes which is based on timeout
+    # If it fails, it will try to delete any existing connection and the reference to the instantiated class
+    # since the retry loop will recreate it
+    def run(self):
+        retry_time = None
+        while True:
+            connection = self._connect()
+            if not connection:
+                continue
```

Comment:
> connection = self._connect()
            if not connection:
                        continue
If for some reason, self._connect() never returns a connection after hitting the retry limit, the loop will continue and try to create the connections again. Should it just raise the exception if it returns None?

Reply:
I think it's fine for the outer loop to continue trying to create/recreate a connection -- the idea is that the microservice container itself will stick around forever, making attempts in case it was a transient network issue getting in the way. To me, that's preferable to letting an exception bubble up and trashing the container.

I'm happy to get into nitty-gritty distributed system strategies (and we probably should in a future PR) but this code was ported over to hoppy from an RRD microservice that worked well enough, so I'm inclined to leave it as-is.

Reply:
I tend to agree with @nanotone. If this microservice can't connect (after it initially connected), the problem is likely with RabbitMQ or the deployment network, so the microservice should continue to retry. 

If the microservice throws an exception and causes the container to fail, then that may distract from our diagnostics. A container failure will cause K8s to continually and needlessly restart the container, consuming resources in the process. If all our microservices followed this convention and RabbitMQ is the problem, they'll all be restarting.

Perhaps this is a good scenario where a Docker health check can be implemented?
@dfitchett Let's capture this concern in a separate ticket for further team discussion.